### PR TITLE
Adds letting ghosts see the hud of the human or borg they observe.

### PR DESCRIPTION
**Pull request by Shadowlight213**
Inspired by goon's observe, but completely my own code because the way they handled it was pretty bad.

<s>Mostly done, remaining changes are fixing the action button hide buttons button appearing for ghosts, with no other buttons visible, and changing its icon and name visible to both human and observer when clicked by the observer.</s>

Finished. Waiting on results of testmerge.

Merge token: #18847

:cl:
add: Ghosts can now see human player's huds when using the OOC -> Observe verb
/:cl:

### CODE REVIEW

**Comment by evsey9** [comment]
So much YES

**Comment by TheNightingale** [comment]
Including cult, rev, gang (etc) icons?

**Comment by Shadowlight213** [comment]
No. This is only for UI stuff.
Sec,med,diag,antag huds are completely different

**Comment by Shadowlight213** [comment]
Should be good now.

**Comment by optimumtact** [comment]
Can you test merge this first, just want to make sure we don't break anything too badly.

**Comment by Shadowlight213** [comment]
@optimumtact I'm planning on that, yes.

**Comment by Shadowlight213** [comment]
tesmerge uncovered bugs.

**Comment by Shadowlight213** [comment]
Going to be gone for 2 weeks with extremely limited internet. I'll keep working on this offline.

**Comment by KorPhaeron** [comment]
Let me know when it's ready to be reopened

**Comment by Armhulen** [comment]
fuck yeah

File: code/_onclick/hud/hud.dm [hunk 1]
```diff
@@ -114,13 +114,18 @@
...
 //Version denotes which style should be displayed. blank or 0 means "next version"
-/datum/hud/proc/show_hud(version = 0)
+/datum/hud/proc/show_hud(version = 0,mob/viewmob)
 	if(!ismob(mymob))
 		return 0
 	if(!mymob.client)
 		return 0
 
-	mymob.client.screen = list()
+	var/mob/screenmob = mymob
```
Comment (RemieRichards) [inline code/_onclick/hud/hud.dm]: ```var/mob/screenmob = viewmob || mymob```
instead of this line and the if.

File: code/_onclick/hud/human.dm [hunk 2]
```diff
@@ -304,102 +314,110 @@
...
 			inv.update_icon()
 
-/datum/hud/human/hidden_inventory_update()
+/datum/hud/human/hidden_inventory_update(mob/viewer)
 	if(!mymob)
 		return
 	var/mob/living/carbon/human/H = mymob
-	if(inventory_shown && hud_shown)
+
+	var/mob/screenmob = H
```
Comment (RemieRichards) [inline code/_onclick/hud/human.dm]: ```var/mob/screenmob = viewer || H```

File: code/_onclick/hud/human.dm [hunk 3]
```diff
@@ -304,102 +314,110 @@
...
+		if(H.wear_mask)	screenmob.client.screen -= H.wear_mask
+		if(H.head)		screenmob.client.screen -= H.head
+
+/datum/hud/human/persistant_inventory_update(mob/viewer)
 	if(!mymob)
 		return
 	var/mob/living/carbon/human/H = mymob
-	if(hud_shown)
+
+	var/mob/screenmob = H
```
Comment (RemieRichards) [inline code/_onclick/hud/human.dm]: ```var/mob/screenmob = viewer || H```

File: code/_onclick/hud/robot.dm [hunk 4]
```diff
@@ -161,20 +181,24 @@
 	r.shown_robot_modules = !r.shown_robot_modules
 	update_robot_modules_display()
 
-/datum/hud/proc/update_robot_modules_display()
+/datum/hud/proc/update_robot_modules_display(mob/viewer)
 	if(!isrobot(mymob)) return
 
 	var/mob/living/silicon/robot/r = mymob
 
+	var/mob/screenmob = r
```
Comment (RemieRichards) [inline code/_onclick/hud/robot.dm]: ```var/mob/screenmob = viewer || r```

File: code/_onclick/hud/robot.dm [hunk 5]
```diff
@@ -206,37 +230,42 @@
...
 
 
-/datum/hud/robot/persistant_inventory_update()
+/datum/hud/robot/persistant_inventory_update(mob/viewer)
 	if(!mymob)
 		return
 	var/mob/living/silicon/robot/R = mymob
-	if(hud_shown)
+
+	var/mob/screenmob = R
```
Comment (RemieRichards) [inline code/_onclick/hud/robot.dm]: ```var/mob/screenmob = viewer || R```

File: code/_onclick/hud/robot.dm [hunk 6]
```diff
@@ -161,20 +181,24 @@
...
+/datum/hud/proc/update_robot_modules_display(mob/viewer)
 	if(!isrobot(mymob)) return
 
 	var/mob/living/silicon/robot/r = mymob
 
+	var/mob/screenmob = r
+	if(viewer)
+		screenmob = viewer
+
 	if(!r.client)
```
Comment (RemieRichards) [inline code/_onclick/hud/robot.dm]: doesn't this need changing now?

**Comment by Shadowlight213** [comment]
This is pretty much finished. Just needs a test merge to triple check, but I think I got all the bugs.

**Comment by jughu** [comment]
Might aswell give ghosts antaghuds at this point

**Comment by Shadowlight213** [comment]
this isn't related to data or antag huds at all.

**Comment by AnturK** [comment]
Seems to add another open empty inventory UI when the target has it open.

**Comment by Shadowlight213** [comment]
@AnturK I'm not sure I understand what you mean? Can you elaborate.

**Comment by AnturK** [comment]
The equipment slots are doubled, namely the ones that can be hidden by the toggle. Seemed to only happen on some people, i guessed it was the differed for toggled off and on.

**Comment by Shadowlight213** [comment]
That's odd. I don't understand why that would happen. Hmm.

**Comment by Shadowlight213** [comment]
Allright. I think that's all the bugs fixed. Empty inventory thing was due to a check missing.
Observer gc will now check if the observer client eye is set to the mob, and remove from the list if the don't match. Previous version had just the client be checked so that logging out wouldn't mess things up.
Also using the observe verb properly calls reset perspective beforehand instead of just setting the eye to the owner mob.

**Comment by optimumtact** [comment]
Have you had another test merge since then? if you can get one with no bugs then I'm happy to merge.

**Comment by optimumtact** [comment]
I'd like to merge this @Shadowlight213

**Comment by Shadowlight213** [comment]
Yah. I just testmerged it.

**Comment by AnturK** [comment]
Is this ready ?

**Comment by Shadowlight213** [comment]
Yes.

**Pull request merged**